test/generators/html/gen_html/gen_html.ml

```diff
@@ -0,0 +1,92 @@
+open Sexplib0.Sexp
+
+type sexp = t = Atom of string | List of sexp list
+
+let read_lines ic =
+  let lines = ref [] in
+  try
+    while true do
+      lines := input_line ic :: !lines
+    done;
+    assert false
+  with End_of_file -> List.rev !lines
+
+let lines_of_command cmd args =
+  let inp = Unix.open_process_in (Filename.quote_command cmd args) in
+  let lines = read_lines inp in
+  let status =
+    match Unix.close_process_in inp with
+    | Unix.WEXITED n -> `Exit n
+    | _ -> `Signal
+  in
+  (lines, status)
+
+let expected_html_targets odocl =
+  lines_of_command "odoc" [ "html-targets"; "-o"; "html.gen"; odocl ]
+
+let tweak_html_target target =
+  String.(concat "." (split_on_char '/' (sub target 14 (length target - 14))))
+
+let html_gen_targets odocl =
+  let targets, status = expected_html_targets odocl in
+  match status with
+  | `Exit 0 ->
+      List.map
+        (fun t ->
+          List
+            [
+              Atom "with-stdout-to";
+              Atom (tweak_html_target t ^ ".gen");
+              List [ Atom "progn"; List [ Atom "system"; Atom ("cat " ^ t) ] ];
+            ])
+        targets
+  (* TODO: enhance error handling *)
+  | _ -> []
+
+let target_rule odocl =
+  List
+    [
+      Atom "rule";
+      List
+        [
+          Atom "action";
+          List
+            (Atom "progn"
+             ::
+             List
+               [
+                 Atom "run";
+                 Atom "odoc";
+                 Atom "html-generate";
+                 Atom "--indent";
+                 Atom "-o";
+                 Atom "html.gen";
+                 Atom ("%{dep:" ^ odocl ^ "}");
+               ]
+             :: html_gen_targets odocl);
+        ];
+    ]
+
+let diff_rule t =
+  List
+    [
+      Atom "rule";
+      List [ Atom "alias"; Atom "runtest" ];
+      List [ Atom "action"; List [ Atom "diff"; Atom t; Atom (t ^ ".gen") ] ];
+    ]
+
+let diff_rules targets =
+  List.iter
+    (fun t -> Printf.printf "%s\n\n" (to_string_hum (diff_rule t)))
+    targets
+
+let gen_html_rule filenames =
+  filenames
+  |> List.iter (fun odocl ->
+         let targets, _ = expected_html_targets odocl in
+         Printf.printf "%s\n\n" (to_string_hum (target_rule odocl));
```

Comment:
`target_rule` will call again `expected_html_targets` indirectly, which isn't good because it's a slow operation.
You should find a way to pass `targets` through this function.

Reply:
good catch!

Reply:
done!